### PR TITLE
PF432 corrige l'apparition des messages d'erreurs inopportuns

### PR DESCRIPTION
**Pull request by qcattez**
Bug lié à l'utilisation de flash au lieu de flash.now quand on render une page au lieu de faire un redirect_to.

Du coup j'ai parcouru le code de l'appli et j'ai corrigé tous les flashs par des flashs.now là où c'était nécessaire.

### CODE REVIEW

**Comment by kemenaran** [comment]
Wow, trop bien – je n'arrivais pas à comprendre d'où venaient ces messages à la con ! ❤️

**Comment by qcattez** [comment]
;)

File: spec/features/1_demarrage/etape1_demarrer_projet_spec.rb [hunk 1]
```diff
@@ -32,6 +32,24 @@
...
+    expect(page).to have_content(I18n.t('demarrage_projet.etape1_demarrage_projet.erreurs.enregistrement_demandeur'))
+    within '.civilite' do choose('Monsieur') end
+    click_button I18n.t('demarrage_projet.action')
+
+    expect(page).to have_current_path projet_avis_impositions_path(projet)
+    expect(page).not_to have_content(I18n.t('demarrage_projet.etape1_demarrage_projet.erreurs.enregistrement_demandeur'))
+    expect(projet.demandeur_principal.civilite).to eq("mr")
+    expect(projet.email).to eq("demandeur@exemple.fr")
+    expect(projet.tel).to eq("01 02 03 04 05")
+  end
```
Comment (kemenaran) [inline spec/features/1_demarrage/etape1_demarrer_projet_spec.rb]: Il n'y aurait pas moyen de faire rentrer ce que teste cette spec dans "quand je rentre des données invalides je vois un message d'erreur" juste en dessus ? Les deux ont l'air très similaires…

Comment (qcattez) [inline spec/features/1_demarrage/etape1_demarrer_projet_spec.rb]: Le truc c'est que c'est un test pour reproduire le bug. Et comme le bug c'est l'apparition d'un message d'erreur sur une autre page là où on ne l'attend pas, ça ne rentre pas dans le scénario "je vois un message d'erreur".
Je peux faire un scénario "reproduction de bug" si tu veux ?

Comment (kemenaran) [inline spec/features/1_demarrage/etape1_demarrer_projet_spec.rb]: Ah, effectivement, je vois l'idée.

Mmm, dans ce cas on pourrait écrire un scénario qui testerait explicitement "Après avoir eu une erreur, quand je soumets une requête valide, le message d'erreur disparait". Ça expliquerait clairement le comportement qu'on teste.

Mais si on part dans cette voie, théoriquement on voudrait un test similaire pour chaque endroit où on utilise `flash.now`… Et je ne pense pas qu'on veuille un test d'acceptance pour à chaque fois que ce edge-case peut théoriquement se produire : c'est trop coûteux, ÀMHA.

Y'aurait pas moyen de tester qu'on utilise `flash.now` (et non pas `flash`) au niveau du contrôleur ? Si y'a moyen, je pense que ça serait une bonne option. Et si y'a pas moyen, on peut reformuler/réduire le test d'acceptance pour indiquer précisément le comportement qu'on teste, et se dire que ça sera déjà bien.

File: spec/controllers/demarrage_projet_controller_spec.rb [hunk 2]
```diff
@@ -156,4 +156,30 @@
...
+
+    before(:each) do
+      projet.demandeur_principal.update_attribute(:civilite, nil)
+      post :etape1_recuperation_infos, params
+      projet.reload
+    end
+
+    it "affiche une erreur" do
+      expect(flash[:alert]).to eq I18n.t('demarrage_projet.etape1_demarrage_projet.erreurs.enregistrement_demandeur')
+    end
```
Comment (kemenaran) [inline spec/controllers/demarrage_projet_controller_spec.rb]: Ok, je vois pourquoi on ne peut pas ré-utiliser les constructions des specs précédentes :)

Mais du coup on n'est pas obligé de se cogner tout le formalisme de `let(:projet_params)`, `let(:params)`, et tout. On peut juste déclarer un `before do`, mettre notre `post` avec les paramètres hardcodés, et basta :)

Comment (qcattez) [inline spec/controllers/demarrage_projet_controller_spec.rb]: Remarques prises en compte ;)

**Comment by qcattez** [comment]
Tous les flashs.now ont été testés manuellement et ils corrigent bien chacun un comportement non souhaité

**Comment by jvignolles** [comment]
Au top ! Merci !
